package/gluon-statistics-mcast/src/respondd.c

```diff
@@ -0,0 +1,478 @@
+	if (!rule_obj)
+		goto err;
+
+	/* sanity check, should not exist yet */
+	ret = json_object_object_get_ex(rule_obj, dirstr[dir], NULL);
+	if (ret)
+		goto err;
+
+	/* json-c < 0.13 */
+	json_object_object_add(rule_obj, dirstr[dir], dir_obj);
+	return 0;
+err:
+	json_object_put(dir_obj);
+	return -ENOMEM;
+}
+
+static int parse_line(struct json_object *obj, enum capdir dir, char *line)
+{
+	int64_t bytes;
+	int64_t packets;
+	char *start;
+	int ret;
+
+	/* skip lines prefixed with '%' */
+	if (line[0] == '%')
+		return 0;
+
+	/* packets */
+	start = strrchr(line, ':');
+	if (!start)
+		return -EINVAL;
+
+	ret = parse_num(&start[1], &packets);
+	if (ret < 0)
+		return ret;
+
+	*start = '\0';
+
+	/* bytes */
+	start = strrchr(line, ':');
+	if (!start)
+		return -EINVAL;
+
+	ret = parse_num(&start[1], &bytes);
+	if (ret < 0)
+		return ret;
+
+	*start = '\0';
+	/* line: is now rule name */
+
+	ret = add_line_to_json(obj, dir, line, &bytes, &packets);
+	if (ret < 0)
+		return -EINVAL;
+
+	return 0;
+}
+
+static int parse_socket(const char *socketname, struct json_object *iface_obj,
+			enum capdir dir)
+{
+	char linebuf[1024];
+	struct sockaddr_un addr;
+	int sd, ret;
+	FILE *file;
+
+	sd = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (sd < 0)
+		return sd;
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sun_family = AF_UNIX;
+	strncpy(addr.sun_path, socketname, sizeof(addr.sun_path) - 1);
+
+	ret = connect(sd, (const struct sockaddr *)&addr,
+		      sizeof(addr));
+	if (ret < 0)
+		goto out;
+
+	file = fdopen(sd, "r");
+	if (!file) {
+		ret = -EACCES;
+		goto out;
+	}
+
+	/* Does this really help or maybe even worsen performance? */
+	setlinebuf(file);
+
+	while (!feof(file)) {
+		if (!fgets(linebuf, sizeof(linebuf)-1, file))
+			break;
+
+		parse_line(iface_obj, dir, linebuf);
+	}
+
+	ret = 0;
+out:
+	close(sd);
+	return ret;
+}
+
+static void parse_socket_subdir(struct dirent *subdir, struct json_object *obj)
+{
+	struct json_object *subdir_obj;
+	struct dirent *dp;
+	DIR *dir;
+
+	if (chdir(subdir->d_name) < 0)
+		return;
+
+	dir = opendir("./");
+	if (!dir)
+		goto err;
+
+	subdir_obj = json_object_new_object();
+	if (!subdir_obj)
+		goto err2;
+
+	json_object_object_add(obj, subdir->d_name, subdir_obj);
+
+	while ((dp = readdir(dir))) {
+		struct json_object *iface_obj;
+		enum capdir capdir;
+
+		if (!strcmp(dp->d_name, ".") || !strcmp(dp->d_name, ".."))
+			continue;
+
+		iface_obj = parse_socket_name(dp->d_name, subdir_obj, &capdir);
+		if (!iface_obj)
+			continue;
+
+		parse_socket(dp->d_name, iface_obj, capdir);
+	}
+
+err2:
+	closedir(dir);
+err:
+	chdir("..");
+}
+
+static struct json_object *get_mcaststats(void)
+{
+	struct json_object *obj;
+	struct dirent *dp;
+	DIR *dir;
+
+	obj = json_object_new_object();
+	if (!obj)
+		return NULL;
+
+	if (chdir(BPFCOUNTD_SOCKS_DIR) < 0)
+		return obj;
+
+	dir = opendir("./");
+	if (!dir)
+		return obj;
+
+	while ((dp = readdir(dir))) {
+		if (!strcmp(dp->d_name, ".") || !strcmp(dp->d_name, ".."))
```

Comment:
Maybe just skip all files starting with `.`?

Reply:
I like that idea! Will change that.